pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -0,0 +1,1139 @@
+use anyhow::Context;
+use pageserver_api::models::{
+    HistoricLayerInfo, LayerAccessKind, LayerResidenceEventReason, LayerResidenceStatus,
+};
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use std::sync::{Arc, Weak};
+use std::time::SystemTime;
+use tracing::Instrument;
+use utils::lsn::Lsn;
+use utils::sync::heavier_once_cell;
+
+use crate::config::PageServerConf;
+use crate::context::RequestContext;
+use crate::repository::Key;
+use crate::tenant::{remote_timeline_client::LayerFileMetadata, RemoteTimelineClient, Timeline};
+
+use super::delta_layer::{self, DeltaEntry};
+use super::image_layer;
+use super::{
+    AsLayerDesc, LayerAccessStats, LayerAccessStatsReset, LayerFileName, PersistentLayerDesc,
+    ValueReconstructResult, ValueReconstructState,
+};
+
+/// A Layer contains all data in a "rectangle" consisting of a range of keys and
+/// range of LSNs.
+///
+/// There are two kinds of layers, in-memory and on-disk layers. In-memory
+/// layers are used to ingest incoming WAL, and provide fast access to the
+/// recent page versions. On-disk layers are stored as files on disk, and are
+/// immutable. This type represents the on-disk kind.
```

Comment:
Maybe a rustdoc link to the in-memory layer would be useful here? e.g.:

```suggestion
/// immutable. This type represents the on-disk kind, while in-memory layers are represented by [`InMemoryLayer`]
```

Reply:
ce9b096f3